rasterio/_warp.pyx

```diff
@@ -360,35 +360,54 @@ def _reproject(
     cdef MemoryDataset src_mem = None
 
     try:
-
         # If the source is an ndarray, we copy to a MEM dataset.
         # We need a src_transform and src_dst in this case. These will
         # be copied to the MEM dataset.
         if dtypes.is_ndarray(source):
             if not src_crs:
                 raise CRSError("Missing src_crs.")
-            if src_nodata is None and hasattr(source, 'fill_value'):
-                # source is a masked array
-                src_nodata = source.fill_value
+
             # ensure data converted to numpy array
-            source = np.asanyarray(source)
+            if hasattr(source, "mask"):
+                source = np.ma.asanyarray(source)
+            else:
+                source = np.asanyarray(source)
+
             # Convert 2D single-band arrays to 3D multi-band.
             if len(source.shape) == 2:
                 source = source.reshape(1, *source.shape)
+
             src_count = source.shape[0]
             src_bidx = range(1, src_count + 1)
-            src_mem = MemoryDataset(source,
-                                         transform=format_transform(src_transform),
-                                         gcps=gcps,
-                                         rpcs=rpcs,
-                                         crs=src_crs,
-                                         copy=True)
+
+            if hasattr(source, "mask"):
```

Comment:
we have twice the `if hasattr(source, "mask")` test, could those block be merged together?

Reply:
@vincentsarago except that the first block needs to run before the input is reshaped and the second block needs to run after the input is reshaped. It's a bit gross, but I'm inclined to let it go for now.